File: Data Structures & Algorithms/minimum-interval-including-query/submission-2.cpp
class Solution {
public:
    vector<int> minInterval(vector<vector<int>>& intervals, vector<int>& queries) {
        sort(intervals.begin(),intervals.end());
        int n = intervals.size();
        map<int,int> mp;
        for(int el:queries){
            mp[el]=-1;
        }
        int i=0;
        auto cmp = [](auto &a,auto &b){
            return a.first>b.first;
        };
        priority_queue<pair<int,int>,vector<pair<int,int>>,decltype(cmp)> pq(cmp);
        for(auto &it:mp){
            int query = it.first;
            while(i<n && intervals[i][0]<=query){
                int a = intervals[i][0];
                int b = intervals[i][1];
                pq.push({(b-a+1),b});
                i++;
            }
            while(!pq.empty() && pq.top().second<query){
                pq.pop();
            }
            if(!pq.empty()){
                it.second=pq.top().first;
            }
        }
        vector<int> ans;
        for(int q:queries){
            ans.push_back(mp[q]);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/asteroid-collision/submission-2.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        ans:List[int] = []
        for ast in asteroids:
            while ans and ans[-1]>0 and ast<0 and ans[-1]<=abs(ast):
                if ans.pop()+ast==0:
                    ast=0
                    break
            if ans and ans[-1]>0 and ast<0 and ans[-1]>abs(ast):
                continue
            if ast!=0:
                ans.append(ast)
        return ans

                    
                
                

